File: ingestion-worker/src/pipeline.ts
/**
 * Document extraction & paragraph-aware chunking (worker side).
 * Same semantics as the web app's pipeline, but uses the singleton Tesseract
 * worker and caps OCR rendering for very large scanned PDFs.
 */
import { getDocument } from 'pdfjs-dist/legacy/build/pdf.mjs';
import { createCanvas, Path2D, DOMMatrix, ImageData } from '@napi-rs/canvas';
import mammoth from 'mammoth';
import { OfficeParser } from 'officeparser';
import { ocrImageBuffer } from './ocr.js';

// pdfjs render path expects these as globals in Node.
function ensureCanvasGlobals() {
  if (!(globalThis as any).Path2D) (globalThis as any).Path2D = Path2D;
  if (!(globalThis as any).DOMMatrix) (globalThis as any).DOMMatrix = DOMMatrix;
  if (!(globalThis as any).ImageData) (globalThis as any).ImageData = ImageData;
}
ensureCanvasGlobals();

const MIN_TEXT_CHARS_PER_PAGE = parseInt(process.env.OCR_MIN_TEXT_CHARS || '20', 10);
const OCR_MAX_PAGES = parseInt(process.env.OCR_MAX_PAGES || '50', 10);

export interface ExtractedPage {
  pageNumber: number;
  text: string;
  isImage?: boolean;
}
export interface ExtractionResult {
  fullText: string;
  pages: ExtractedPage[];
}
export interface DocumentChunk {
  chunkIndex: number;
  totalChunks: number;
  rawContent: string;
  pageStart: number;
  pageEnd: number;
  paragraphId?: string;
  chunkType: 'text' | 'image' | 'table';
  charStart?: number;
  charEnd?: number;
  fileName: string;
  title?: string;
  stream?: string;
  semester?: string;
  section?: string;
  subject?: string;
  module?: string;
}

async function renderPdfPageToPng(doc: any, pageNum: number, scale = 2): Promise<Buffer | null> {
  try {
    const page = await doc.getPage(pageNum);
    const viewport = page.getViewport({ scale });
    const canvas = createCanvas(viewport.width, viewport.height);
    const ctx = canvas.getContext('2d');
    ctx.fillStyle = '#ffffff';
    ctx.fillRect(0, 0, viewport.width, viewport.height);
    await page.render({ canvasContext: ctx, viewport } as any).promise;
    return canvas.toBuffer('image/png');
  } catch (err: any) {
    console.warn(`PDF page ${pageNum} render failed (OCR unavailable):`, err.message);
    return null;
  }
}

export async function extractPdf(buffer: Buffer): Promise<ExtractionResult> {
  const pages: ExtractedPage[] = [];
  let doc: any = null;
  try {
    doc = await getDocument({
      data: new Uint8Array(buffer),
      useSystemFonts: true,
      disableFontFace: true,
    } as any).promise;

    let ocrRendered = 0;
    for (let pageNum = 1; pageNum <= doc.numPages; pageNum++) {
      let text = '';
      try {
        const page = await doc.getPage(pageNum);
        const tc = await page.getTextContent();
        let lastY: number | null = null;
        const parts: string[] = [];
        for (const item of tc.items as any[]) {
          const y = item.transform?.[5];
          if (lastY !== null && y !== undefined && Math.abs(y - lastY) > 2) parts.push('\n');
          parts.push(item.str ?? '');
          lastY = y ?? lastY;
        }
        text = parts.join('').trim();
      } catch (e: any) {
        console.warn(`PDF page ${pageNum} text extraction error:`, e.message);
      }

      if (text.length < MIN_TEXT_CHARS_PER_PAGE && ocrRendered < OCR_MAX_PAGES) {
        const png = await renderPdfPageToPng(doc, pageNum);
        if (png) {
          const ocrText = await ocrImageBuffer(png);
          ocrRendered++;
          if (ocrText.length > text.length) {
            pages.push({ pageNumber: pageNum, text: ocrText, isImage: true });
            continue;
          }
        }
        pages.push({ pageNumber: pageNum, text, isImage: text.length === 0 });
      } else {
        pages.push({ pageNumber: pageNum, text });
      }
    }
  } catch (err: any) {
    console.warn('PDF parsing failed, falling back to whole-buffer OCR:', err.message);
    const ocrText = await ocrImageBuffer(buffer);
    if (ocrText) pages.push({ pageNumber: 1, text: ocrText, isImage: true });
  } finally {
    if (doc) {
      try { await doc.destroy(); } catch {}
    }
  }
  const fullText = pages.map((p) => p.text).join('\n\n');
  return { fullText, pages: pages.length ? pages : [{ pageNumber: 1, text: fullText }] };
}

export async function extractDocx(buffer: Buffer): Promise<ExtractionResult> {
  const result = await mammoth.extractRawText({ buffer });
  const text = (result.value || '').trim();
  return { fullText: text, pages: [{ pageNumber: 1, text }] };
}

export async function extractPptx(buffer: Buffer): Promise<ExtractionResult> {
  try {
    const ast = await OfficeParser.parseOffice(buffer, { fileType: 'pptx' } as any);
    const text = (typeof ast.toText === 'function' ? ast.toText() : '') || '';
    return { fullText: text, pages: [{ pageNumber: 1, text: text.trim() }] };
  } catch (err: any) {
    console.error('PPTX extraction error:', err.message);
    return { fullText: '', pages: [] };
  }
}

export async function runOcrOnBuffer(buffer: Buffer): Promise<ExtractionResult> {
  const text = await ocrImageBuffer(buffer);
  return { fullText: text, pages: [{ pageNumber: 1, text, isImage: true }] };
}

export function extractMarkdown(buffer: Buffer): ExtractionResult {
  let text = buffer.toString('utf-8').replace(/^---\s*\n[\s\S]*?\n---\s*\n/, '').trim();
  return { fullText: text, pages: [{ pageNumber: 1, text }] };
}

export function extractPlainText(buffer: Buffer): ExtractionResult {
  const text = buffer.toString('utf-8').trim();
  return { fullText: text, pages: [{ pageNumber: 1, text }] };
}

export async function extractDocumentContent(
  fileName: string,
  buffer: Buffer,
  mimeType?: string
): Promise<ExtractionResult> {
  const ext = fileName.split('.').pop()?.toLowerCase() || '';
  const mime = (mimeType || '').toLowerCase();
  if (mime.includes('pdf') || ext === 'pdf') return extractPdf(buffer);
  if (mime.includes('presentation') || ext === 'pptx') return extractPptx(buffer);
  if (mime.includes('wordprocessingml') || mime.includes('msword') || ext === 'docx' || ext === 'doc')
    return extractDocx(buffer);
  if (mime.includes('image') || ['png','jpg','jpeg','webp','gif','bmp','tiff'].includes(ext))
    return runOcrOnBuffer(buffer);
  if (ext === 'md' || ext === 'markdown' || mime.includes('markdown')) return extractMarkdown(buffer);
  return extractPlainText(buffer);
}

function splitParagraphs(text: string): Array<{ text: string; start: number }> {
  const out: Array<{ text: string; start: number }> = [];
  const re = /\n[ \t]*\n/g;
  let start = 0;
  let match: RegExpExecArray | null;
  while (start < text.length) {
    re.lastIndex = start;
    match = re.exec(text);
    const end = match ? match.index : text.length;
    const para = text.slice(start, end).trim();
    if (para) out.push({ text: para, start });
    if (!match) break;
    start = match.index + match[0].length;
  }
  return out;
}

export function chunkExtractedDocument({
  extraction,
  fileName,
  title,
  stream,
  semester,
  section,
  subject,
  module,
  chunkSize = parseInt(process.env.CHUNK_SIZE || '500', 10),
  chunkOverlap = parseInt(process.env.CHUNK_OVERLAP || '50', 10),
}: {
  extraction: ExtractionResult;
  fileName: string;
  title?: string;
  stream?: string;
  semester?: string;
  section?: string;
  subject?: string;
  module?: string;
  chunkSize?: number;
  chunkOverlap?: number;
}): DocumentChunk[] {
  const rawChunks: Array<{
    text: string;
    pageStart: number;
    pageEnd: number;
    paragraphId?: string;
    chunkType: 'text' | 'image' | 'table';
    charStart?: number;
    charEnd?: number;
  }> = [];

  for (const page of extraction.pages) {
    const pageText = page.text.trim();
    if (!pageText) continue;
    if (page.isImage) {
      rawChunks.push({
        text: pageText,
        pageStart: page.pageNumber,
        pageEnd: page.pageNumber,
        paragraphId: `${page.pageNumber}:img`,
        chunkType: 'image',
      });
      continue;
    }
    const paragraphs = splitParagraphs(pageText);
    let paraIdx = 0;
    for (const para of paragraphs) {
      paraIdx++;
      const paragraphId = `${page.pageNumber}:${paraIdx}`;
      let start = 0;
      while (start < para.text.length) {
        const end = Math.min(start + chunkSize, para.text.length);
        const chunkText = para.text.slice(start, end).trim();
        if (chunkText.length > 20) {
          rawChunks.push({
            text: chunkText,
            pageStart: page.pageNumber,
            pageEnd: page.pageNumber,
            paragraphId,
            chunkType: 'text',
            charStart: para.start + start,
            charEnd: para.start + end,
          });
        }
        start += chunkSize - chunkOverlap;
        if (start >= para.text.length) break;
      }
    }
  }

  if (rawChunks.length === 0 && extraction.fullText.trim()) {
    const text = extraction.fullText.trim();
    let start = 0;
    while (start < text.length) {
      const end = Math.min(start + chunkSize, text.length);
      const chunkText = text.slice(start, end).trim();
      if (chunkText.length > 20) {
        rawChunks.push({ text: chunkText, pageStart: 1, pageEnd: 1, paragraphId: '1:1', chunkType: 'text', charStart: start, charEnd: end });
      }
      start += chunkSize - chunkOverlap;
    }
  }

  const total = rawChunks.length;
  return rawChunks.map((c, idx) => ({
    chunkIndex: idx,
    totalChunks: total,
    rawContent: c.text,
    pageStart: c.pageStart,
    pageEnd: c.pageEnd,
    paragraphId: c.paragraphId,
    chunkType: c.chunkType,
    charStart: c.charStart,
    charEnd: c.charEnd,
    fileName,
    title: title || fileName,
    stream: stream || 'General',
    semester: semester || 'General',
    section: section || 'General',
    subject: subject || 'General',
    module: module || 'General',
  }));
}

File: optimized-worker/worker/pipeline.py
"""
Document extraction and storage download.

Uses PyMuPDF (fitz) for PDF extraction — significantly faster than pdfjs-dist.
Supports: PDF, DOCX, PPTX, MD, TXT, images (OCR).
"""
from __future__ import annotations

import io
import logging
import time
from dataclasses import dataclass, field
from typing import Optional

logger = logging.getLogger(__name__)


@dataclass
class ExtractedPage:
    page_number: int
    text: str
    is_image: bool = False


@dataclass
class ExtractionResult:
    full_text: str
    pages: list[ExtractedPage] = field(default_factory=list)
    timings: dict = field(default_factory=dict)


async def extract_pdf(buffer: bytes) -> ExtractionResult:
    """Extract text from PDF using PyMuPDF with per-page OCR fallback."""
    import fitz  # PyMuPDF
    from .ocr import ocr_pil_image
    from .config import get_settings
    from PIL import Image

    settings = get_settings()
    min_chars = settings.ocr_min_text_chars
    max_ocr = settings.ocr_max_pages

    pages: list[ExtractedPage] = []
    timings = {}

    t0 = time.perf_counter()
    try:
        doc = fitz.open(stream=buffer, filetype="pdf")
        timings["pdf_open_s"] = round(time.perf_counter() - t0, 4)

        ocr_count = 0
        t_text = time.perf_counter()

        for page_num in range(len(doc)):
            page = doc[page_num]
            text = page.get_text("text").strip()

            if len(text) < min_chars and ocr_count < max_ocr:
                # Render to image for OCR
                t_ocr = time.perf_counter()
                pix = page.get_pixmap(dpi=200)
                img = Image.frombytes("RGB", [pix.width, pix.height], pix.samples)
                ocr_text = ocr_pil_image(img)
                ocr_count += 1

                if len(ocr_text) > len(text):
                    pages.append(ExtractedPage(
                        page_number=page_num + 1, text=ocr_text, is_image=True
                    ))
                    continue

            pages.append(ExtractedPage(
                page_number=page_num + 1, text=text, is_image=len(text) == 0
            ))

        timings["text_extraction_s"] = round(time.perf_counter() - t_text, 4)
        timings["ocr_pages"] = ocr_count
        doc.close()

    except Exception as e:
        logger.warning("PDF parsing failed, attempting whole-buffer OCR: %s", e)
        from .ocr import ocr_image_buffer
        ocr_text = ocr_image_buffer(buffer)
        if ocr_text:
            pages.append(ExtractedPage(page_number=1, text=ocr_text, is_image=True))

    full_text = "\n\n".join(p.text for p in pages)
    if not pages:
        pages = [ExtractedPage(page_number=1, text=full_text)]

    return ExtractionResult(full_text=full_text, pages=pages, timings=timings)


async def extract_docx(buffer: bytes) -> ExtractionResult:
    """Extract text from DOCX using mammoth."""
    import mammoth

    t0 = time.perf_counter()
    result = mammoth.extract_raw_text(io.BytesIO(buffer))
    text = (result.value or "").strip()
    elapsed = time.perf_counter() - t0

    return ExtractionResult(
        full_text=text,
        pages=[ExtractedPage(page_number=1, text=text)],
        timings={"docx_extraction_s": round(elapsed, 4)},
    )


async def extract_pptx(buffer: bytes) -> ExtractionResult:
    """Extract text from PPTX using python-pptx."""
    from pptx import Presentation

    t0 = time.perf_counter()
    prs = Presentation(io.BytesIO(buffer))
    texts = []
    for slide_num, slide in enumerate(prs.slides, 1):
        slide_text_parts = []
        for shape in slide.shapes:
            if shape.has_text_frame:
                for paragraph in shape.text_frame.paragraphs:
                    slide_text_parts.append(paragraph.text)
        texts.append("\n".join(slide_text_parts))

    full_text = "\n\n".join(texts)
    elapsed = time.perf_counter() - t0

    return ExtractionResult(
        full_text=full_text,
        pages=[ExtractedPage(page_number=1, text=full_text)],
        timings={"pptx_extraction_s": round(elapsed, 4)},
    )


async def extract_markdown(buffer: bytes) -> ExtractionResult:
    """Extract text from Markdown, stripping YAML frontmatter."""
    import re

    text = buffer.decode("utf-8", errors="replace")
    text = re.sub(r"^---\s*\n[\s\S]*?\n---\s*\n", "", text).strip()
    return ExtractionResult(
        full_text=text,
        pages=[ExtractedPage(page_number=1, text=text)],
    )


async def extract_plain_text(buffer: bytes) -> ExtractionResult:
    """Extract plain text."""
    text = buffer.decode("utf-8", errors="replace").strip()
    return ExtractionResult(
        full_text=text,
        pages=[ExtractedPage(page_number=1, text=text)],
    )


async def extract_image(buffer: bytes) -> ExtractionResult:
    """OCR an image."""
    from .ocr import ocr_image_buffer

    t0 = time.perf_counter()
    text = ocr_image_buffer(buffer)
    elapsed = time.perf_counter() - t0

    return ExtractionResult(
        full_text=text,
        pages=[ExtractedPage(page_number=1, text=text, is_image=True)],
        timings={"ocr_s": round(elapsed, 4)},
    )


async def extract_document(
    file_name: str, buffer: bytes, mime_type: str = ""
) -> ExtractionResult:
    """Route extraction based on file extension/MIME type."""
    ext = file_name.rsplit(".", 1)[-1].lower() if "." in file_name else ""
    mime = (mime_type or "").lower()

    if mime.startswith("application/pdf") or ext == "pdf":
        return await extract_pdf(buffer)
    if "presentation" in mime or ext == "pptx":
        return await extract_pptx(buffer)
    if "wordprocessingml" in mime or "msword" in mime or ext in ("docx", "doc"):
        return await extract_docx(buffer)
    if mime.startswith("image/") or ext in ("png", "jpg", "jpeg", "webp", "gif", "bmp", "tiff"):
        return await extract_image(buffer)
    if ext in ("md", "markdown") or "markdown" in mime:
        return await extract_markdown(buffer)

    return await extract_plain_text(buffer)
